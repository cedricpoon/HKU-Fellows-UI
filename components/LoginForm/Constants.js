export const LOGO_SIZE = 150;
