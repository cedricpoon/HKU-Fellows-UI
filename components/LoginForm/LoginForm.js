import React, { Component } from 'react';
import { Platform, Image, Dimensions, Keyboard } from 'react-native';
import { Form, Item, Label, Input, Text, Button, Icon, View } from 'native-base';
import PropTypes from 'prop-types';

import background from './background';
import styles from './Styles';
import { localize } from 'hkufui/locale';

const locale = localize({ language: 'en', country: 'hk' });

export class LoginForm extends Component {
  constructor(props) {
    super(props);

    this.state = { validity: { username: true, password: true } };
    this._setValidity = this._setValidity.bind(this);
    this._handleChangeWrapper = this._handleChangeWrapper.bind(this);
    this._validate = this._validate.bind(this);
  }

  _setValidity(name, isValid, more = {}) {
    this.setState(prevState => ({
      ...more,
      validity: {
        ...prevState.validity,
        [name]: isValid
      }
    }));
  }

  _validate() {
    const { onLogin, alert } = this.props;
    const { username, password } = this.state;

    Keyboard.dismiss();

    let valid = true;
    if (!username) {
      this._setValidity('username', false);
      valid = false;
    }
    if (!password) {
      this._setValidity('password', false);
      valid = false;
    }
    if (valid) {
      onLogin({ username, password });
    } else {
      alert(locale['login.missingParam']);
    }
  }

  _handleChangeWrapper(name) {
    return (newValue) => {
      this._setValidity(name, true, { [name]: newValue });
    }
  }

  render() {
    const { validity } = this.state;
    const { loggingIn } = this.props;

    const deviceHeight = Dimensions.get('window').height;
    const deviceWidth = Dimensions.get('window').width;

    return (
      <View style={styles.fullPageContainer}>
        <View style={styles.inputForm}>
          <Form>
            <Item floatingLabel error={!validity.username}>
              <Label><Text note style={styles.placeholder}>{locale['login.username']}</Text></Label>
              <Input
                autoCorrect={false}
                keyboardType={Platform.OS === 'android' ? 'email-address' : 'ascii-capable'}
                onChangeText={this._handleChangeWrapper('username')}
              />
              {!validity.username &&
                <Icon name='exclamationcircleo' type='AntDesign' />
              }
            </Item>
            <Item floatingLabel error={!validity.password}>
              <Label><Text note style={styles.placeholder}>{locale['login.password']}</Text></Label>
              <Input
                secureTextEntry
                onChangeText={this._handleChangeWrapper('password')}
              />
              {!validity.password &&
                <Icon name='exclamationcircleo' type='AntDesign' />
              }
            </Item>
            <Item style={styles.noBorderItem}>
              <Label>
                <Text style={[styles.declaration, styles.placeholder]}>
                  {locale['login.declaration']}
                </Text>
              </Label>
            </Item>
          </Form>
          <Button block iconRight transparent dark
            style={styles.submit}
            onPressIn={this._validate}
            disabled={loggingIn}
          >
            <Text>{locale['login.button']}</Text>
            <Icon name='login' type='AntDesign' />
          </Button>
        </View>
        <Image
          source={{ uri: background }}
          style={[
            { width: deviceWidth, height: deviceHeight },
            styles.backdrop
          ]}
          resizeMode='repeat'
        />
      </View>
    );
  }
}

LoginForm.propTypes = {
  alert: PropTypes.func.isRequired,
  onLogin: PropTypes.func.isRequired,
  loggingIn: PropTypes.bool
}

export default LoginForm;
