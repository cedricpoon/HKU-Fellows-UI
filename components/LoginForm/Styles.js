import { StyleSheet } from 'react-native';

import { logo } from 'hkufui/theme/palette';
import { gridBase as base } from 'hkufui/theme/grid';
import { fontSize } from 'hkufui/theme/grid'

export default StyleSheet.create({
  fullPageContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  inputForm: {
    paddingHorizontal: base * 6,
    paddingBottom: base * 6,
    zIndex: 2,
    alignItems: 'center'
  },
  fullWidth: {
    width: '100%'
  },
  submit: {
    marginTop: base * 6,
    height: base * 9
  },
  declaration: {
    fontSize: fontSize.miniSize,
    fontStyle: 'italic'
  },
  placeholder: {
    color: logo.black
  },
  noBorderItem: {
    borderBottomWidth: 0,
    marginTop: base * 3
  },
  backdrop: {
    zIndex: 1
  }
});
