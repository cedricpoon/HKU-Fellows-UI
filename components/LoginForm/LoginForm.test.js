import React from 'react';
import { shallow } from 'enzyme';
import LoginForm from './LoginForm';

describe('Testing LoginForm Component', () => {
  const requiredProps = {
    alert: () => {},
    onLogin: () => {}
  };

  it('renders as expected', () => {
    const wrapper = shallow(
      <LoginForm {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('is logging in', () => {
    const wrapper = shallow(
      <LoginForm {...requiredProps} loggingIn />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
