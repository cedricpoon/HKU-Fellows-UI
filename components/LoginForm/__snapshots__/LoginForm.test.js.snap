// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginForm Component is logging in 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Styled(ViewNB)
    style={
      Object {
        "alignItems": "center",
        "paddingBottom": 30,
        "paddingHorizontal": 30,
        "zIndex": 2,
      }
    }
  >
    <Logo
      size={150}
    />
    <Styled(Form)
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Styled(Item)
        error={false}
        floatingLabel={true}
      >
        <Styled(Label)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "color": "#4E4A47",
              }
            }
          >
            HKU Portal UID
          </Styled(Text)>
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="ascii-capable"
          onChangeText={[Function]}
        />
      </Styled(Item)>
      <Styled(Item)
        error={false}
        floatingLabel={true}
      >
        <Styled(Label)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "color": "#4E4A47",
              }
            }
          >
            HKU Portal PIN
          </Styled(Text)>
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          secureTextEntry={true}
        />
      </Styled(Item)>
      <Styled(Item)
        style={
          Object {
            "borderBottomWidth": 0,
            "marginTop": 15,
          }
        }
      >
        <Styled(Label)>
          <Styled(Text)
            style={
              Array [
                Object {
                  "fontSize": 10,
                  "fontStyle": "italic",
                },
                Object {
                  "color": "#4E4A47",
                },
              ]
            }
          >
            **Your PIN will NOT be saved in our database
          </Styled(Text)>
        </Styled(Label)>
      </Styled(Item)>
    </Styled(Form)>
    <Styled(Spinner)
      inverse={true}
      size="small"
      style={
        Object {
          "height": 45,
          "marginTop": 30,
        }
      }
    />
  </Styled(ViewNB)>
  <Backdrop
    height={1334}
    style={
      Object {
        "zIndex": 1,
      }
    }
    width={750}
  />
</KeyboardAvoidingView>
`;

exports[`Testing LoginForm Component renders as expected 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Styled(ViewNB)
    style={
      Object {
        "alignItems": "center",
        "paddingBottom": 30,
        "paddingHorizontal": 30,
        "zIndex": 2,
      }
    }
  >
    <Logo
      size={150}
    />
    <Styled(Form)
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Styled(Item)
        error={false}
        floatingLabel={true}
      >
        <Styled(Label)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "color": "#4E4A47",
              }
            }
          >
            HKU Portal UID
          </Styled(Text)>
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="ascii-capable"
          onChangeText={[Function]}
        />
      </Styled(Item)>
      <Styled(Item)
        error={false}
        floatingLabel={true}
      >
        <Styled(Label)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "color": "#4E4A47",
              }
            }
          >
            HKU Portal PIN
          </Styled(Text)>
        </Styled(Label)>
        <Styled(Input)
          onChangeText={[Function]}
          secureTextEntry={true}
        />
      </Styled(Item)>
      <Styled(Item)
        style={
          Object {
            "borderBottomWidth": 0,
            "marginTop": 15,
          }
        }
      >
        <Styled(Label)>
          <Styled(Text)
            style={
              Array [
                Object {
                  "fontSize": 10,
                  "fontStyle": "italic",
                },
                Object {
                  "color": "#4E4A47",
                },
              ]
            }
          >
            **Your PIN will NOT be saved in our database
          </Styled(Text)>
        </Styled(Label)>
      </Styled(Item)>
    </Styled(Form)>
    <Styled(Button)
      block={true}
      dark={true}
      iconRight={true}
      onPress={[Function]}
      style={
        Object {
          "height": 45,
          "marginTop": 30,
        }
      }
      transparent={true}
    >
      <Styled(Text)>
        Login
      </Styled(Text)>
      <Styled(Icon)
        name="login"
        type="AntDesign"
      />
    </Styled(Button)>
  </Styled(ViewNB)>
  <Backdrop
    height={1334}
    style={
      Object {
        "zIndex": 1,
      }
    }
    width={750}
  />
</KeyboardAvoidingView>
`;
